src/librustc_lint/types.rs

```diff
@@ -427,13 +427,24 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
                                               adding a member to this struct");
                         }
 
-                        // We can't completely trust repr(C) markings; make sure the
-                        // fields are actually safe.
+                        // We can't completely trust repr(C) and repr(transparent) markings;
+                        // make sure the fields are actually safe.
                         let mut all_phantom = true;
                         for field in &def.struct_variant().fields {
                             let field_ty = cx.fully_normalize_associated_types_in(
                                 &field.ty(cx, substs)
                             );
+                            // repr(transparent) types are allowed to have arbitrary ZSTs, not just
+                            // PhantomData -- skip checking all ZST fields
```

Comment:
We *could* restrict it to `PhantomData`, didn't know the lint was special-casing that already.

Reply:
I'm torn. Supporting only `PhantomData` would be the more conservative option and I don't have a use case for other ZSTs at hand. OTOH `repr(transparent)` is explicitly ignoring all ZSTs, so it makes sense to allow them for C interop too even though `repr(C)` warns about other ZSTs.

Reply:
Ah. The RFC says all ZSTs are fine, and that seems nicer to me than special casing PhantomData, but being conservative and following the precedent of the improper-ctypes lint also makes sense. If anyone want to start out accepting only PhantomData, I'll implement that. Otherwise, well, it's not exactly a burden to support other ZSTs.

Reply:
Wouldn't allowing all ZSTs could change the alignment, making it not really transparent?

```rust
#[repr(transparent)]
struct U8 {
    a: u8,
    not_really: [u32; 0],
}
```

Reply:
@kennytm A good point, but this was already brought up in the RFC discussion, with the proposed solution (that I implemented) being to prohibit ZSTs with alignment > 1.

Reply:
@rkuppe thanks, found the check in `check_transparent`